File: go/internal/percept/message.go
package percept

import "context"

// Role identifies who authored a Message.
type Role int

const (
	RoleUser Role = iota
	RoleAssistant
)

// Message is one turn in a conversation - the value-object shape Model
// needs, independent of Event's identity/audit concerns.
type Message struct {
	Role    Role
	Content string
}

// Model turns a conversation into a streamed reply - the domain's core
// capability, mechanism-agnostic. Reply returns as soon as streaming has
// started (not once it's finished); chunks arrive on the returned
// channel as they're produced, and the channel is closed when the reply
// is complete. Mid-stream errors are out of scope - only a failure
// before streaming starts is reported via the returned error.
type Model interface {
	Reply(ctx context.Context, messages []Message) (<-chan string, error)
}

// ToMessages converts the transcript into the form Model expects.
func ToMessages(events []Event) []Message {
	messages := make([]Message, len(events))
	for i, e := range events {
		role := RoleUser
		if e.Sender == SenderAssistant {
			role = RoleAssistant
		}
		messages[i] = Message{Role: role, Content: e.Content}
	}
	return messages
}

File: rust/src/percept/message.rs
use super::{Event, Sender};

/// Identifies who authored a Message.
#[derive(Clone, Copy, PartialEq, Eq)]
pub enum Role {
    User,
    Assistant,
}

/// One turn in a conversation - the value-object shape Model needs,
/// independent of Event's identity/audit concerns.
pub struct Message {
    pub role: Role,
    pub content: String,
}

/// Turns a conversation into a reply - the domain's core capability,
/// mechanism-agnostic. The method itself stays synchronous; the caller
/// runs it off the UI thread (e.g. via `spawn_blocking`) instead.
/// `Send + Sync` are required so a provider can be shared (`Arc`)
/// across that boundary.
pub trait Model: Send + Sync {
    fn reply(&self, messages: &[Message]) -> Result<String, Box<dyn std::error::Error>>;
}

/// Converts the transcript into the form Model expects.
pub fn to_messages(events: &[Event]) -> Vec<Message> {
    events
        .iter()
        .map(|e| Message {
            role: match e.sender {
                Sender::User => Role::User,
                Sender::Assistant => Role::Assistant,
            },
            content: e.content.clone(),
        })
        .collect()
}
